File: sdk/go/knowledge.go
package contextos

import (
	"context"
	"encoding/json"
	"fmt"
)

// KnowledgeEntity represents a knowledge entity.
type KnowledgeEntity struct {
	ID         string            `json:"id"`
	Name       string            `json:"name"`
	Type       string            `json:"type"`
	Properties map[string]string `json:"properties,omitempty"`
	CreatedAt  string            `json:"created_at"`
}

// KnowledgeRelationship represents a knowledge relationship.
type KnowledgeRelationship struct {
	ID         string            `json:"id"`
	SourceID   string            `json:"source_id"`
	TargetID   string            `json:"target_id"`
	Type       string            `json:"type"`
	Properties map[string]string `json:"properties,omitempty"`
	Weight     float64           `json:"weight"`
}

// KnowledgeQueryRequest is the request to query knowledge.
type KnowledgeQueryRequest struct {
	EntityName string `json:"entity_name,omitempty"`
 EntityType string `json:"entity_type,omitempty"`
	MaxDepth  int    `json:"max_depth,omitempty"`
}

// KnowledgeQueryResponse is the response from knowledge query.
type KnowledgeQueryResponse struct {
	Entities     []KnowledgeEntity     `json:"entities"`
	Relationships []KnowledgeRelationship `json:"relationships"`
}

// KnowledgeService handles knowledge operations.
type KnowledgeService struct {
	client *Client
}

// AddEntity adds a knowledge entity.
func (s *KnowledgeService) AddEntity(ctx context.Context, entity *KnowledgeEntity) (*KnowledgeEntity, error) {
	var result KnowledgeEntity
	body, err := s.client.doRequest(ctx, "POST", "/api/v1/knowledge/entities", entity)
	if err != nil {
		return nil, err
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}
	return &result, nil
}

// AddRelationship adds a knowledge relationship.
func (s *KnowledgeService) AddRelationship(ctx context.Context, rel *KnowledgeRelationship) (*KnowledgeRelationship, error) {
	var result KnowledgeRelationship
	body, err := s.client.doRequest(ctx, "POST", "/api/v1/knowledge/relationships", rel)
	if err != nil {
		return nil, err
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}
	return &result, nil
}

// Query queries the knowledge graph.
func (s *KnowledgeService) Query(ctx context.Context, req *KnowledgeQueryRequest) (*KnowledgeQueryResponse, error) {
	var result KnowledgeQueryResponse
	body, err := s.client.doRequest(ctx, "POST", "/api/v1/knowledge/query", req)
	if err != nil {
		return nil, err
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}
	return &result, nil
}

// GetEntity gets a knowledge entity by name.
func (s *KnowledgeService) GetEntity(ctx context.Context, name string) (*KnowledgeEntity, error) {
	var result KnowledgeEntity
	body, err := s.client.doRequest(ctx, "GET", "/api/v1/knowledge/entities/"+name, nil)
	if err != nil {
		return nil, err
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}
	return &result, nil
}

File: sdk/rust/src/knowledge.rs
use serde::{Deserialize, Serialize};

use crate::error::Result;

/// Knowledge entity.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeEntity {
    pub id: String,
    pub name: String,
    #[serde(rename = "type")]
    pub entity_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub properties: Option<std::collections::HashMap<String, String>>,
    pub created_at: String,
}

/// Knowledge relationship.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeRelationship {
    pub id: String,
    pub source_id: String,
    pub target_id: String,
    #[serde(rename = "type")]
    pub relationship_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub properties: Option<std::collections::HashMap<String, String>>,
    pub weight: f64,
}

/// Knowledge query request.
#[derive(Debug, Serialize)]
pub struct KnowledgeQueryRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub entity_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub entity_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_depth: Option<u32>,
}

/// Knowledge query response.
#[derive(Debug, Deserialize)]
pub struct KnowledgeQueryResponse {
    pub entities: Vec<KnowledgeEntity>,
    pub relationships: Vec<KnowledgeRelationship>,
}

/// Knowledge service.
pub struct KnowledgeService {
    base_url: String,
    api_key: String,
}

impl KnowledgeService {
    pub(crate) fn new(base_url: String, api_key: String) -> Self {
        KnowledgeService { base_url, api_key }
    }

    /// Add a knowledge entity.
    pub async fn add_entity(&self, entity: &KnowledgeEntity) -> Result<KnowledgeEntity> {
        let url = format!("{}/api/v1/knowledge/entities", self.base_url);
        let resp = reqwest::Client::new()
            .post(&url)
            .header("Authorization", format!("Bearer {}", self.api_key))
            .json(entity)
            .send()
            .await
            .map_err(|e| crate::error::Error::Request(e.to_string()))?;

        let result: KnowledgeEntity = resp
            .json()
            .await
            .map_err(|e| crate::error::Error::Serialization(e.to_string()))?;

        Ok(result)
    }

    /// Query the knowledge graph.
    pub async fn query(&self, entity_name: Option<&str>, max_depth: Option<u32>) -> Result<KnowledgeQueryResponse> {
        let req = KnowledgeQueryRequest {
            entity_name: entity_name.map(|s| s.to_string()),
            entity_type: None,
            max_depth,
        };

        let url = format!("{}/api/v1/knowledge/query", self.base_url);
        let resp = reqwest::Client::new()
            .post(&url)
            .header("Authorization", format!("Bearer {}", self.api_key))
            .json(&req)
            .send()
            .await
            .map_err(|e| crate::error::Error::Request(e.to_string()))?;

        let result: KnowledgeQueryResponse = resp
            .json()
            .await
            .map_err(|e| crate::error::Error::Serialization(e.to_string()))?;

        Ok(result)
    }
}
